schema.cc

```diff
@@ -159,6 +169,20 @@ static const dht::sharder& get_sharder(unsigned shard_count, unsigned ignore_msb
     return *it->second;
 }
 
+static std::shared_ptr<repair_history_map> get_repair_history_map(utils::UUID id) {
+    auto it = repair_history_maps.find(id);
+    if (it != repair_history_maps.end()) {
+        return it->second;
+    } else {
+        repair_history_maps[id] = std::make_shared<repair_history_map>();
+        return repair_history_maps[id];
+    }
+}
```

Comment:
This is using a secret global to inject the repair_history_map into schema. It works, but it's a hack. I'd rather have the data propagate directly.

Reply:
Avi, it is easy to say let's pass get_gc_before function everywhere. It is very painful and ugly. 

schema  object already has everything else we need to calculate gc_before, e.g., the mode and the parameters. 

We also have things like 

```
thread_local std::map<sstring, std::unique_ptr<dht::i_partitioner>> partitioners;
thread_local std::map<std::pair<unsigned, unsigned>, std::unique_ptr<dht::sharder>> sharders;

std::reference_wrapper<const dht::i_partitioner> _partitioner;
std::reference_wrapper<const dht::sharder> _sharder;
```

Reply:
It's true that we already attach sharders to schema, and as predicted here we have a bug related to this discovered some time ago: https://github.com/scylladb/scylla/issues/7642

Reply:
Looking at #7642, unlike_sharder, there are no different settings for  _repair_history_map. It is just a pointer.

Reply:
We can't keep on piling hacks. It makes this series more difficult, yes, but it doesn't create a gooball that someone will have to untangle later.

Reply:
I do not understand why you think it is hack. I agree it is not nice to have a global. But I do not think it is a hack. It is pretty natrual to me that we put something that is related to gc_before. 

@avikivity you are welcome to suggest better ideas (other than passing the get_gc_before function from the table object everywhere) than what I proposed here.

Reply:
scylla.git$ git grep  'thread_local'|wc -l
447

Reply:
Btw, once we elimiate 

static thread_local schema_registry registry;

we can pass the repair history maps from the main.cc along with the registry.

When registry creates a schema, we use the maps from main.cc to construct the schema.

No more global repair_histroy_maps. 

E.g., schema_registry_entry::get_schema -> frozen_schema::unfreeze -> db::schema_tables::create_table_from_mutations.

Reply:
> Looking at #7642, unlike_sharder, there are no different settings for _repair_history_map. It is just a pointer.

Yes. It's still a problem if we allow the user of schema to set the pointer, since we don't want the same pointer on a different shard. Or maybe you intend the getter inside schema to always use the thread_local and not allow the user to set the pointer? Then you don't need to go through the schema object, you can access the thread local directly.

Reply:
> Btw, once we elimiate
> 
> static thread_local schema_registry registry;

I have a PR doing just that, preparatory part here: https://github.com/scylladb/scylla/pull/9635 (original Pr that was too big: https://github.com/scylladb/scylla/pull/9620)

Reply:
> Yes. It's still a problem if we allow the user of schema to set the pointer, since we don't want the same pointer on a different shard. Or maybe you intend the getter inside schema to always use the thread_local and not allow the user to set the pointer? Then you don't need to go through the schema object, you can access the thread local directly.

As you can see in v6, there is no setter at all.  Accessing the thread local directly means a global function like: get_local_repair_history_map().

This is possible.But i do not like this global function.  I believe Avi hate it too. 

Another reason that I think it makes  sense to put into schema is:  Getting the repair history map itself is not enough, we still need something from the schema to calculate gc_before, e..g, the tombstone_gc_ mode and propagation delay. 

Why putting it into multiple places instead of one place.

Reply:
> I have a PR doing just that, preparatory part here: #9635 (original Pr that was too big: #9620)

Cool. Thanks. As the time being, I won't depend on this though.

Reply:
Another thing I want to mention is that if we chose to pass the get_gc_before function around, we probably need to pass more than one function. We need another one for giving the answer for a specific sstable:

get_gc_before_for_a_range(range, query_time)

in addition to the current one

get_gc_before(dk, query_time)

Reply:
can we make it a virtual class? then we can have a specialized impl for each mode.

Reply:
The schema controls the tombstone gc options mode. The repair history map is part of the gc options mode = repair. From this point of view, attaching the history map to the schema actually makes sense.

Schema is not just about describing the structure of the table. We already has plenty of such items like paxos_grace_seconds(), gc_grace_seconds(), min_compaction_threshold, etc, which describe the property of the table.

Updating repair history does not change the structure of the table, but it changes the property of a table.

Anyway, I do not mind to have a global function like below for now like you suggested:

gc_clock::time_point get_gc_before(schema_ptr s, const dht::decorated_key& dk, const gc_clock::time_point& query_time) const;

@avikivity @tgrabiec please ack

(This will make it harder to get rid of the global than attaching get_gc_before to schema (mentioned here: https://github.com/scylladb/scylla/pull/9219#discussion_r751269174). It is easy to move back if we want.)

So, let's move on. We have more important things to worry about for this series.

Reply:
I would still rather see the `gc_before` propagated to compactors. Most places should have a table object nearby, or they are in tests, which can use a dummy gc_before, like we do with the `never_gc` can_gc_fn.

Reply:
> Schema is not just about describing the structure of the table. We already has plenty of such items like paxos_grace_seconds(), gc_grace_seconds(), min_compaction_threshold, etc, which describe the property of the table.
> 
> Updating repair history does not change the structure of the table, but it changes the property of a table.
> 

Yes, we also put properties there. But the point about value semantics stands, the repair history would not be part of the schema value (so not included in the digest, shared among instances), whereas the properties are.

Reply:
It is easy to say we have table object nearby. If you look more, there are many cases, it is in the middle of no where and huge amount of code need to be changed.  If you really want to pass it everywhere, we can do it later. And you are welcome to send a patch. 

In v6, I am moving everything out of schema.hh. I hope this is less controversial so we can move forward.